### PR TITLE
Add ED25519 Signature demo.

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: demos/signature/EVP_ED_Signature_demo.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        fprintf(stderr, "EVP_DigestVerifyInit_ex failed.\n");
+        goto cleanup;
+    }
+    /*
+     * ED25519 only supports the one shot interface using EVP_DigestVerify()
+     * The streaming EVP_DigestVerifyUpdate() API is not supported.
+     */
+    if (!EVP_DigestVerify(verify_context, sig_value, sig_len,
+                          tbs, tbs_len)) {
+        fprintf(stderr, "EVP_DigestVerifyUpdate(hamlet_1) failed.\n");
```
Comment (t8m) [inline demos/signature/EVP_ED_Signature_demo.c:118]: Replace the error text with `EVP_DigestVerify failed.\n`

**Review by t8m — CHANGES_REQUESTED** [review]

File: demos/signature/EVP_ED_Signature_demo.c [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        fprintf(stderr, "EVP_DigestSignInit_ex failed.\n");
+        goto cleanup;
+    }
+
+    /* Calculate the required size for the signature by passing a NULL buffer. */
+    if (!EVP_DigestSign(sign_context, NULL, &sig_len, tbs, tbs_len)) {
+        fprintf(stderr, "EVP_DigestSign using NULL buffer failed.\n");
+        goto cleanup;
+    }
+    sig_value = OPENSSL_zalloc(sig_len);
```
Comment (mattcaswell) [inline demos/signature/EVP_ED_Signature_demo.c:65]: Why zalloc here and not malloc?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: demos/signature/EVP_ED_Signature_demo.c (hunk 2 above)
Comment (slontis) [reply demos/signature/EVP_ED_Signature_demo.c:65]: I guess it is a choice between having a zero'd value and something potentially random when it goes wrong.

File: demos/signature/EVP_ED_Signature_demo.c (hunk 1 above)
Comment (slontis) [reply demos/signature/EVP_ED_Signature_demo.c]: thanks

**Comment by slontis** [comment]
I created this as I can see that there have been multiple questions asked about ED verification (both here and in stack overflow)

File: demos/README.txt [hunk 3]
```diff
@@ -51,9 +51,11 @@ pkread.c               Print out a description of a PKCS12 file.
 pkwrite.c              Add a password to an existing PKCS12 file.
 
 signature:
-EVP_Signature_demo.c   Compute and verify a signature from multiple buffers
-rsa_pss_direct.c       Compute and verify an RSA-PSS signature from a hash
-rsa_pss_hash.c         Compute and verify an RSA-PSS signature over a buffer
+EVP_EC_Signature_demo.c   Compute and verify a EC signature.
```
Comment (tom-cosgrove-arm) [inline demos/README.txt:54]: (minor) "an" EC signature

File: demos/signature/EVP_ED_Signature_demo.c [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    /* Create a signature context */
+    sign_context = EVP_MD_CTX_new();
+    if (sign_context == NULL) {
+        fprintf(stderr, "EVP_MD_CTX_new failed.\n");
+        goto cleanup;
+    }
+
+    /*
+     * Initialize the sign context using a ED25519 private key
```
Comment (tom-cosgrove-arm) [inline demos/signature/EVP_ED_Signature_demo.c:47]: (minor) "an" ED25519 private key

File: demos/README.txt [hunk 5]
```diff
@@ -51,9 +51,11 @@ pkread.c               Print out a description of a PKCS12 file.
...
+EVP_DSA_Signature_demo.c  Compute and verify a DSA signature.
+EVP_ED_Signature_demo.c   Compute and verify a ED25519 signature.
```
Comment (mattcaswell) [inline demos/README.txt:56]: While we're at it: "an" ED25519 signature

**Review by mattcaswell — APPROVED** [review]
Approved with or without the minor typos noted above being corrected

**Comment by slontis** [comment]
Stupid English language :)

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM, thanks for fixing those nits - yes, English is bizarre!

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master. Thank you.

**Pull request closed without merging**